### PR TITLE
convert expr.func is xyz to isinstance(expr,xyz)

### PR DESCRIPTION
**Pull request by nilabja-bhattacharya**
Fixes to issues as mentioned in issue #13111 and as per discussion in issue #13342

### CODE REVIEW

File: sympy/core/expr.py [hunk 1]
```diff
@@ -2256,7 +2256,7 @@ def extract_branch_factor(self, allow_half=False):
         args = Mul.make_args(self)
         exps = []
         for arg in args:
-            if arg.func is exp_polar:
+			if isinstance(arg,exp_polar):
```
Comment (ylemkimon) [inline sympy/core/expr.py]: Use 4 spaces instead of tab and space after comma. (Same for below)

File: sympy/core/mod.py [hunk 2]
```diff
@@ -93,14 +93,14 @@ def doit(p, q):
...
-        if p.func is cls:
+		if isinstance(p,cls):
             qinner = p.args[1]
             if qinner % q == 0:
                 return cls(p.args[0], q)
             elif (qinner*(q - qinner)).is_nonnegative:
                 # |qinner| < |q| and have same sign
                 return p
-        elif (-p).func is cls:
+		elif isinstance((-p),cls):
```
Comment (ylemkimon) [inline sympy/core/mod.py]: Parenthesis is not necessary for function arguments.

**Review by ylemkimon — CHANGES_REQUESTED** [review]

File: sympy/core/power.py [hunk 3]
```diff
@@ -212,11 +212,11 @@ def __new__(cls, b, e, evaluate=None):
                 return S.One
             else:
                 # recognize base as E
-                if not e.is_Atom and b is not S.Exp1 and b.func is not exp_polar:
+                if not e.is_Atom and b is not S.Expl and not(instance(b,exp_polar)):
```
Comment (ylemkimon) [inline sympy/core/power.py]: It is `Exp1`(the number one, not alphabet l) and did you mean `not isinstance(b, exp_polar)`?

File: sympy/core/mod.py [hunk 4]
```diff
@@ -93,14 +93,14 @@ def doit(p, q):
...
-        if p.func is cls:
+        if isinstance(p, cls):
             qinner = p.args[1]
             if qinner % q == 0:
                 return cls(p.args[0], q)
             elif (qinner*(q - qinner)).is_nonnegative:
                 # |qinner| < |q| and have same sign
                 return p
-        elif (-p).func is cls:
+        elif isinstance((-p), cls):
```
Comment (ylemkimon) [inline sympy/core/mod.py]: You can do `isinstance(-p, cls)`.

**Review by ylemkimon — CHANGES_REQUESTED** [review]

**Comment by gxyd** [comment]
You can try to fix all the `is` comparison issues in this pull request. You can do `git grep -E 'func is'` in SymPy root folder to get a list of (possibly) all such issues.

**Comment by nilabja10201992** [comment]
@gxyd Why are Travis checks failing?

**Comment by gxyd** [comment]
Don't worry about that. They failed on the past few builds, but on the latest build tests seems to be passing. Here they, passed :)

**Comment by nilabja10201992** [comment]
@gxyd I'll make changes to all files where changes can be made (there are lots of files where it can be implemented), thank you for guiding.

**Comment by ylemkimon** [comment]
I think `x.func is type` in the tests can be left as is.

**Comment by nilabja10201992** [comment]
@ylemkimon okay I'll make changes only if the file isn't in test folder of sympy.

File: sympy/functions/elementary/trigonometric.py [hunk 5]
```diff
@@ -1305,30 +1305,30 @@ def eval(cls, arg):
...
             x = arg.args[0]
             return 1 / x
 
-        if arg.func is atan2:
+        if isinstance(arg, atan2):
             y, x = arg.args
             return x/y
 
-        if arg.func is asin:
+        if isinstance(arg,asin):
```
Comment (ylemkimon) [inline sympy/functions/elementary/trigonometric.py]: Space after comma

File: sympy/functions/special/error_functions.py [hunk 6]
```diff
@@ -778,12 +778,12 @@ def eval(cls, z):
         elif z is S.One:
             return S.Infinity
 
-        if (z.func is erf) and z.args[0].is_real:
+        if (isinstance(z, erf)) and z.args[0].is_real:
```
Comment (ylemkimon) [inline sympy/functions/special/error_functions.py]: The parenthesis surrounding isinstance is unnecessary. Same for below.

**Review by ylemkimon — CHANGES_REQUESTED** [review]

File: sympy/functions/special/error_functions.py [hunk 7]
```diff
@@ -778,12 +778,12 @@ def eval(cls, z):
...
             return S.Infinity
 
-        if (z.func is erf) and z.args[0].is_real:
+        if isinstance(z, erf) and z.args[0].is_real:
             return z.args[0]
 
         # Try to pull out factors of -1
         nz = z.extract_multiplicatively(-1)
-        if nz is not None and ((nz.func is erf) and (nz.args[0]).is_real):
+        if nz is not None and ((isinstance(nz, erf)) and (nz.args[0]).is_real):
```
Comment (ylemkimon) [inline sympy/functions/special/error_functions.py]: The parenthesis surrounding isinstance is unnecessary.

**Comment by ylemkimon** [comment]
There seems some `expr.func is type` left.
```python
sympy/concrete/expr_with_limits.py:27:    if hasattr(function, 'func') and function.func is Equality:
sympy/core/power.py:641:        if old.func is self.func and self.exp == old.exp:
sympy/core/power.py:646:        if old.func is self.func and self.base == old.base:
sympy/functions/special/zeta_functions.py:163:        if z.func is exp and (z.args[0]/(pi*I)).is_Rational or z in [-1, I, -I]:
sympy/integrals/deltafunctions.py:55:        if arg.is_Pow and arg.base.func is DiracDelta:
sympy/integrals/deltafunctions.py:67:            elif arg.is_Pow and arg.base.func is DiracDelta:
sympy/integrals/integrals.py:773:        if f.func is Piecewise:
sympy/integrals/risch.py:306:            if (i.base.is_Pow or i.base.func is exp and i.exp.is_Rational)]
sympy/logic/algorithms/dpll.py:283:                P, value = sym, not (literal.func is Not)
sympy/logic/boolalg.py:1069:    if info[0].func is info[2]:
sympy/logic/boolalg.py:1071:            if arg.func is info[1]:
sympy/logic/boolalg.py:1079:    elif info[0].func is info[1]:
sympy/logic/boolalg.py:1272:    if expr.func is function2:
sympy/logic/boolalg.py:1287:    if expr.func is not function1:
sympy/logic/boolalg.py:1296:        elif cls.func is not function2:
sympy/series/gruntz.py:278:        if e.args[0].func is log:
sympy/simplify/fu.py:326:                any(any(ai.func is cos or ai.is_Pow and ai.base is cos
sympy/simplify/fu.py:382:                    k.args[1].func is cos:
sympy/simplify/fu.py:1482:        if not (isinstance(rv, Pow) and rv.base.func is sin):
sympy/simplify/fu.py:1511:        if not (isinstance(rv, Pow) and rv.base.func is cos):
sympy/simplify/fu.py:1544:        if rv.base.func is tan:
sympy/simplify/fu.py:1546:        elif rv.base.func is sin:
sympy/simplify/fu.py:1548:        elif rv.base.func is cos:
sympy/simplify/fu.py:1931:        if c.func is not s.func:
sympy/simplify/fu.py:1937:                if not ((ca.func is sa.func) is (cb.func is sb.func)):
sympy/simplify/fu.py:1942:                return gcd, n1, n2, ca.args[0], sa.args[0], ca.func is sa.func
sympy/simplify/powsimp.py:678:        if glogb.args[0].is_Pow or glogb.args[0].func is exp:
sympy/solvers/ode.py:2273:                if newi.lhs.func is log and newi.rhs == 0:
sympy/solvers/ode.py:2656:        if expr.func is Add:
sympy/solvers/ode.py:2678:    if expr.func is Equality:
sympy/solvers/ode.py:2681:        if lhs.func is Symbol and lhs in Cs:
sympy/solvers/ode.py:2814:                if m.func is exp:
sympy/solvers/ode.py:2817:                    infac = any(fi.func is exp for t in m.args
```

Thank you!

**Comment by nilabja10201992** [comment]
@ylemkimon the ode.py file is commented there is no code where it can be implemented. And in case of fu.py the statements are like rv.base.func is tan, if I change it to isinstance(rv.base, tan) will it cause some unnecessary  problem? I'll look into that and make the changes and I'll wait for your reply.

File: sympy/integrals/deltafunctions.py [hunk 8]
```diff
@@ -55,14 +55,14 @@ def change_mul(node, x):
...
             dirac = arg
         else:
             new_args.append(arg)
     if not dirac:  # there was no simple dirac
         new_args = []
         for arg in sorted_args:
-            if arg.func is DiracDelta:
+            if isinstance(arg, DiracDelta):
                 new_args.append(arg.expand(diracdelta=True, wrt=x))
             elif arg.is_Pow and arg.base.func is DiracDelta:
```
Comment (smichr) [inline sympy/integrals/deltafunctions.py]: It is perhaps worth checking again for `.func is ` to see if any others (like this one) got missed.

File: sympy/logic/boolalg.py [hunk 9]
```diff
@@ -1271,7 +1271,7 @@ def _is_form(expr, function1, function2):
     # Special case of a single expression of function2
     if expr.func is function2:
```
Comment (smichr) [inline sympy/logic/boolalg.py]: here's another that got missed

File: sympy/logic/boolalg.py [hunk 10]
```diff
@@ -1290,13 +1290,13 @@ def _is_form(expr, function1, function2):
     for cls in expr.args:
         if cls.is_Atom:
             continue
-        if cls.func is Not:
+        if isinstance(cls, Not):
             if not cls.args[0].is_Atom:
                 return False
         elif cls.func is not function2:
```
Comment (smichr) [inline sympy/logic/boolalg.py]: another

File: sympy/series/gruntz.py [hunk 11]
```diff
@@ -269,10 +269,10 @@ def mrv(e, x):
...
             return s, expr**e
-    elif e.func is log:
+    elif isinstance(e, log):
         s, expr = mrv(e.args[0], x)
         return s, log(expr)
-    elif e.func is exp:
+    elif isinstance(e, exp):
         # We know from the theory of this algorithm that exp(log(...)) may always
         # be simplified here, and doing so is vital for termination.
         if e.args[0].func is log:
```
Comment (smichr) [inline sympy/series/gruntz.py]: another

File: sympy/simplify/fu.py [hunk 12]
```diff
@@ -1922,15 +1922,15 @@ def pow_cos_sin(a, two):
...
 
     # check them
-    if (not ca) and cb or ca and ca.func is sin:
+    if (not ca) and cb or ca and isinstance(ca, sin):
         coa, ca, sa, cob, cb, sb = cob, cb, sb, coa, ca, sa
         n1, n2 = n2, n1
     if not two:  # need cos(x) and cos(y) or sin(x) and sin(y)
         c = ca or sa
         s = cb or sb
         if c.func is not s.func:
```
Comment (smichr) [inline sympy/simplify/fu.py]: another (but this one should be `if not isinstance(c.func, s.func)`

File: sympy/simplify/powsimp.py [hunk 13]
```diff
@@ -674,7 +674,7 @@ def nc_gcd(aa, bb):
             glogb = _keep_coeff(cg, rg*Add(*[a/g for a in args]))
 
     # now put the log back together again
-    if glogb.func is log or not glogb.is_Mul:
+    if isinstance(glogb, log) or not glogb.is_Mul:
         if glogb.args[0].is_Pow or glogb.args[0].func is exp:
```
Comment (smichr) [inline sympy/simplify/powsimp.py]: another

**Comment by ylemkimon** [comment]
I missed the comments, sorry. `rv.base.func is tan` and similar can be safely converted to `isinstance(rv.base, tan)` since it is equivalent to `(rv.base).func`.

**Comment by nilabja10201992** [comment]
@ylemkimon @smichr` if not ((ca.func is sa.func) is (cb.func is sb.func)):`
will `if not isinstance(isinstacnce(ca, sa.func), isinstance(cb,sb.func)):` be a solution to reduce this?

**Comment by smichr** [comment]
> if not (isinstacnce(ca, sa.func) is isinstance(cb,sb.func)):

That should work, but I would write it with the `not` in the middle:

```
if isinstance(ca, sa.func) is not isinstance(cb, sb.func):
```

**Comment by ylemkimon** [comment]
There seems one more at [sympy/functions/special/zeta_functions.py#L163](https://github.com/nilabja10201992/sympy/blob/2fa80c56468cb7647ec6731a7c5bdb14cb5dde4a/sympy/functions/special/zeta_functions.py#L163) and some at `ode.py`:
```
sympy/solvers/ode.py:2656:        if expr.func is Add:
sympy/solvers/ode.py:2678:    if expr.func is Equality:
sympy/solvers/ode.py:2681:        if lhs.func is Symbol and lhs in Cs:
sympy/solvers/ode.py:2814:                if m.func is exp:
sympy/solvers/ode.py:2817:                    infac = any(fi.func is exp for t in m.args
```

**Comment by smichr** [comment]
Thanks! This is in. Hopefully this won't break too many PRs but this is something that we've said that should be done. So...no better time than today!

**Comment by smichr** [comment]
Thanks, @ylemkimon , for the printouts of where more of these were.

**Comment by nilabja10201992** [comment]
@gxyd @ylemkimon @smichr please suggest some issue that I can work on, "easy fixes" in issues section is too outdated. Thank You.

**Pull request merged**